### ISSUE TITLE
cleanup V3.5 Token-based Session Management

### ISSUE THREAD

**Issue by elarlang** [OWASP/ASVS#1917]
Started to re-investigate issue #1790, then checked the section "V3.5 Token-based Session Management" and reached to the conclusion, that the entire section should be just cleaned up.

At the moment collecting first impressions as feedback. If needed, I split each topic into separate issues.

## TLDR

* Requirements 3.5.3, 3.5.4, 3.5.5, and 3.5.6 are general token validation requirements, that are not limited to session context and should be moved to "V13 API and Web Service"
* Requirement 3.5.1 - it is not a session requirement, should belong to OAuth or be moved together with 3.5.7.
* Requirement 3.5.7 - it is not a session requirement, and should be moved to access control

## 3.5.1

| # | Description | L1 | L2 | L3 | CWE | [NIST &sect;](https://pages.nist.gov/800-63-3/sp800-63b.html) |
| :---: | :--- | :---: | :---: | :---: | :---: | :---: |
| **3.5.1** | [GRAMMAR] Verify that the application allows users to revoke OAuth tokens that form trust relationships with linked applications. | | ✓ | ✓ | 290 | 7.1.2 |

Is it really a session management issue? Seems more like a business logical permission management - if a user has given permission for the application to access some 3rd party application on his/her behalf, then:

  * 1) the application must provide the functionality to revoke those tokens
  * 2) the 3rd party application must revoke them (but it's out of scope from the application point of view)

CWE is not helpful here.

Also, if it is OAuth specific, it should be moved there.

In case it is not OAuth specific, it should be considered as access control requirement and move together with 3.5.7.

## 3.5.3, 3.5.5

| # | Description | L1 | L2 | L3 | CWE | [NIST &sect;](https://pages.nist.gov/800-63-3/sp800-63b.html) |
| :---: | :--- | :---: | :---: | :---: | :---: | :---: |
| **3.5.3** | [MODIFIED, LEVEL L2 > L1] Verify that stateless session tokens make use of a digital signature to protect against tampering and this is checked before processing it further. | ✓ | ✓ | ✓ | 345 | |
| **3.5.5** | [ADDED] Verify that only allow-listed signing algorithms are allowed for a stateless token. | ✓ | ✓ | ✓ | 757 | |

3.5.3 should not be limited with "stateless session tokens", just "stateless tokens".

Not a session specific requirements.

## 3.5.4

| # | Description | L1 | L2 | L3 | CWE | [NIST &sect;](https://pages.nist.gov/800-63-3/sp800-63b.html) |
| :---: | :--- | :---: | :---: | :---: | :---: | :---: |
| **3.5.4** | [ADDED] Verify that stateless tokens are checked for expiration before processing them further. | ✓ | ✓ | ✓ | 613 | |

So it should cover "time window" when the token is valid.

To check the expiration, we need to require `exp` to be present and set.

Not sure what security problem it takes down, but technically it should also cover to check `nbf` (Not Before).

Related RFC https://datatracker.ietf.org/doc/html/rfc7519

Not a session specific requirement.

## 3.5.6

| # | Description | L1 | L2 | L3 | CWE | [NIST &sect;](https://pages.nist.gov/800-63-3/sp800-63b.html) |
| :---: | :--- | :---: | :---: | :---: | :---: | :---: |
| **3.5.6** | [ADDED] Verify that other, security-sensitive attributes of a stateless token are being verified. For example, in a JWT this may include issuer, subject, and audience. | ✓ | ✓ | ✓ | 287 | |

Not a session specific requirement.

## 3.5.7

| # | Description | L1 | L2 | L3 | CWE | [NIST &sect;](https://pages.nist.gov/800-63-3/sp800-63b.html) |
| :---: | :--- | :---: | :---: | :---: | :---: | :---: |
| **3.5.7** | [ADDED] Verify that all active stateless tokens, which are being relied upon for access control decisions, are revoked when admins change the entitlements or roles of the user. | ✓ | ✓ | ✓ | 613 | |

The context here - first-party party tokens. It should be clarified in the requirement if we'll keep it.

Another problem - to have the requirement in the session category, it assumes, that stateless tokens are used as a replacement for stateful session.

As I wrote here: https://github.com/OWASP/ASVS/issues/1790#issuecomment-1823972356

From [NIST SP 800-63B version 4 (not released)](https://pages.nist.gov/800-63-4/sp800-63b.html#access-tokens) 7.1.2 Access Tokens:

> An access token — such as found in OAuth — is used to allow an application to access a set of services on a subscriber’s behalf following an authentication event. The presence of an OAuth access token SHALL NOT be interpreted by the RP as presence of the subscriber, in the absence of other signals. The OAuth access token, and any associated refresh tokens, MAY be valid long after the authentication session has ended and the subscriber has left the application.

If we remove the session context from the requirement, it is still a valid requirement - but should be moved to the "V4 Access Control" category.

**Comment by tghosth** [comment]
3.5.1 - Agree it should be part of the OAuth Chapter
3.5.7 - Agree it can be moved to access control.

The rest, I would support some rewording of the requirements and the section name but I am not convinced about moving them to an entirely different chapter as people will **mostly** (admittedly not always) come across these tokens in a session management context.

**Comment by elarlang** [comment]
Let's try to get the first 2 sorted...

## 3.5.1

Thinking more about 3.5.1 - I think it is not the best fit for OAuth, although suits there better than in the session category, but maybe we should look at it as general access management logic, not as an OAuth protocol-specific requirement.

## 3.5.7

We don't have a matching requirement for the same situation in the "classical session management" - if the permissions for the user are changed, it must have immediate effect. Such as in case those are buffered to session data, those also must be "revoked".

So maybe we need to make this requirement to cover both situations and to be more abstract. Or do we need to have another requirement for "classical session management".


## both

Till we don't have anything better, I propose moving them to the current V4.3 Other Access Control Considerations, or, if anyone has a good proposal, to a new section with the meaning "Access Control Management". Need to re-check when we going to re-work V4.

ping @tghosth

**Comment by tghosth** [comment]
> Thinking more about 3.5.1 - I think it is not the best fit for OAuth, although suits there better than in the session category, but maybe we should look at it as general access management logic, not as an OAuth protocol-specific requirement.

I agree with the concept where access tokens have been issued and I agree with moving it to access control but you would need to reword it to make it a little clearer as to why.

> So maybe we need to make this requirement to cover both situations and to be more abstract. Or do we need to have another requirement for "classical session management".

So the reason this exists is that in a stateless session, the session token might be relied upon for the permissions the user has whereas in stateful it will always be checked against the database. I agree that it could be combined into one requirement and made more abstract

**Comment by randomstuff** [comment]
> Verify that stateless session tokens make use of a digital signature to protect against tampering and this is checked before processing it further.

To be strict this should be "digital signature or MAC": "Verify that stateless session tokens are protected against tampering (eg. using a digital signature or a MAC) and this protection is checked before processing it further."

**Comment by randomstuff** [comment]
> Verify that all active stateless tokens, which are being relied upon for access control decisions, are revoked when admins change the entitlements or roles of the user.

We can't revoke the token if it is really stateless, though.

**Comment by jmanico** [comment]
What is the use case for very short-lived tokens that last just a few minutes? I'm sincerely asking because I have not encountered that before.

**Comment by randomstuff** [comment]
I would say the use case for access tokens which last a few minutes is to have really stateless access tokens while providing short grant revocation propagation.

[RFC6819](https://www.rfc-editor.org/rfc/rfc6819#section-5.1.5.2) says:

> tokens may expire after a few minutes (e.g., for payment transactions) or stay valid for hours (e.g., read access to contacts).

Some examples in the wild:

* [Some Stack Overflow answer](https://stackoverflow.com/a/49556417/637866) suggests that Rakuten access token  is (was?) 5 minutes.
* [Pro Santé Connect](https://industriels.esante.gouv.fr/produits-et-services/pro-sante-connect/documentation-technique) (OIDC) access token is 2 minutes.

**Comment by tghosth** [comment]
I think this issue is waiting for the V3 rework which I don't think has been assigned yet but the points will certainly be considered when it happens

**Comment by jmanico** [comment]
I am worried that 3.5.3 only states that a digital signature should be used to protect against tampering. Many of the IdP's only support RSA or maybe EcDSA with are both very problematic.

The best digital signature today is Edwards Curve digital signatures. Ed25519 is truly what we should all be moving to today. And in the future, we should be considering DSA's like : Dilithium and Falcon which are quantum resistant.

3.5.3 | [MODIFIED, LEVEL L2 > L1] Verify that stateless session tokens make use of a digital signature to protect against tampering and this is checked before processing it further. | ✓ | ✓ | ✓ | 345
-- | -- | -- | -- | -- | --

Even if we just say to use a GOOD digital signature for 3.5.3 would help. The RSA/EcDSA standard of the day are weak, at best. I am of the opinion that we do not want general requirements, we want to be very specific when we can, and this especially goes for cryptography.

**Comment by jmanico** [comment]
Before we  move 3.5.7 I want to make a comment.

3.5.7 | [ADDED] Verify that all active stateless tokens, which are being relied upon for access control decisions, are revoked when admins change the entitlements or roles of the user. | ✓ | ✓ | ✓ | 613
-- | -- | -- | -- | -- | --

Revoking the token when the admin entitlement is not the best strategy. Just like using REDIS for revocation, I can set a REDIS rule to over-ride the tokens access control with the new entitlement and not force a user to log in again, if this is used for session management. Token revocation here is a fairly weak and user-unfriendly method. So if we keep 3.5.7 I'd suggest:

3.5.7 | [ADDED] Verify that all active stateless tokens, which are being relied upon for access control decisions, are revoked or overridden with the new access control entitlement when admins change the entitlements or roles of the user. | ✓ | ✓ | ✓ | 613
-- | -- | -- | -- | -- | --

**Comment by randomstuff** [comment]
> And although this may mean by a strict definition the architecture is not completely stateless, these revocation strategies are still mostly stateless and not not require a full-on server-side sessions.

I think that "stateless" token should be stateless by definition :wink:. If you don't intend to mean that the stateless token is stateless the name should be changed for something more precise.

**Comment by jmanico** [comment]
I don’t believe that strict definition of stateless is useful. Revocation strategies are critical to any kind of token technology. That’s my take on stateless.

**Comment by jmanico** [comment]
Basically, I think of cache blacklisting, token counters and similar as a “minimally statefull approach” that is still very acceptable for stateless technologies and stateless tokens and doesn’t require us to redefine what a stateless token is.

**Comment by elarlang** [comment]
> I don’t believe that strict definition of stateless is useful. 

First URL-encoding was not really a URL-encoding and now stateless is not actually stateless? What's next? :)

By principle, stateless is meant to be stateless. This is the one and only definition for that. https://restfulapi.net/#1-3-stateless

If the requirement talks about something other than a stateless token, we should make the correction to the requirement, not saying "No no no my dear friend, our stateless token may have a state".

The misusage of stateless tokens in the session management context is/was the main reason why I proposed to move this part away from the session management paragraph.

> Revocation strategies are critical to any kind of token technology. That’s my take on stateless.

It is critical, but the question is - how we should define and call such tokens? One way or another you need to "replace the state mechanism" and invent the wheel, which always raising the question - was/is it worth it?

**Comment by jmanico** [comment]
> First URL-encoding was not really a URL-encoding and now stateless is not actually stateless? What's next? :)

Someday when you try to URL encode a JWT or binary data on a URL and realize base64url encoding use cases matter, you will say “OMG Jim was partially correct!.” It’s then when I’ll expect you to apologize. I don’t expect this for 3-10 years. But after that I will ask you to refer to me as your “URL Encoding Master” like from star wars(tm). 😘 
> By principle, stateless is meant to be stateless. This is the one and only definition for that. https://restfulapi.net/#1-3-stateless

“Mostly stateless” is good enough. Access control, override rules and revocation are really necessary in nature systems for session management and similar. Eventually, you’re gonna need to revoke a refresh token in the world of OAuth2 as well.

**Comment by elarlang** [comment]
> It’s then when I’ll expect you to apologize

A bit unreal, no? All I can do is a conference talk about different defense methods 101 - what and in which step of data processing you should use. With an extra - how people without the ability or will to analyze technical processes kill the project.

As no explanation has been analyzed so far, I can not see the point of explaining it further.

Special note: I have never touched binary data in the given topic, it was path traversal, you just try to save your * with this demagogy.

> Mostly stateless

Sounds like "mostly boolean true", or "75% true"? It is stateless or it is not stateless.

To leave the offtopic aside - it's clear that those terms need some serious work and do not cause confusion.

@randomstuff - do you have proposals in mind for that?

**Comment by jmanico** [comment]
I’m making a point that revocation strategies for stateless mechanisms is still quite reasonable. It’s a fair point.

Please Elar, act professionally.

**Comment by jmanico** [comment]
In real-world scenarios, systems *frequently* mix stateless and stateful design elements to find a good balance between security, performance, and scalability. 

One of the common strategies is cache-based revocation so the performance impact is less:

1. Better Security:  Revocation helps reduce the risk of compromised tokens being used until they expire.
2. Trade-offs: This approach adds some state management, which can impact scalability and performance, especially in systems that are spread out over many servers.
3. Implementation: Managing the cache efficiently, such as using distributed caches like Redis or Memcach (very common revocation strategies), can help reduce some of the performance issues.

**Comment by randomstuff** [comment]
"HMAC" should be replaces by "MAC": any other (non-broken) MAC could be used.

(Would it be useful, to mention authenticated encryption algorithms here?)

**Comment by randomstuff** [comment]
"before processing further" is somewhat vague/ambiguous.

**Comment by jmanico** [comment]
So how about:
How about this?

3.5.3 | [MODIFIED, LEVEL L2 > L1] Verify that stateless tokens use a digital signature or MAC to protect against tampering, ensuring it is checked before performing any other actions based on the token's contents. | ✓ | ✓ | ✓ | 345
-- | -- | -- | -- | -- | --

It's a LOT more specific, but too much so?

**Comment by elarlang** [comment]
Each requirement has now a separate issue for discussion and not to be discussed here:

* 3.5.1 - will be moved/replaced to OAuth via #2120
* 3.5.2 - (already moved away)
* 3.5.3 - #2184
* 3.5.4 - #2185
* 3.5.5 - no change proposals made, option to merge into 3.5.3 (#2184)
* 3.5.6 - #1967
* 3.5.7 - will be moved/replaced to V4 via #2059

---

The section name "Token-based Session Management" is just misleading (the same way as "Cookie-based Session Management"). Token or Cookie is just a technical solution, how the "state" is delivered to the server, it's nothing about the management.

Session is a logic on topic of technology. I think session paragraph should not contain sections for validating stateless tokens or cookie setup, as those are technical things to validate also without the session context. In case we move them away, the link/reference must be clear.

**Comment by randomstuff** [comment]
> **Verify that the token is an access token** intended for the service (API). For JWTs this can be achieved by validating the claims 'aud' and 'typ'. If audience is present, it must be validated to match against an allow list for the given API. If the typ claim is present, it should be equal to 'at-jwt'. (see new issue https://github.com/OWASP/ASVS/issues/2363)

In this formulation, I read:

> Verify that the token is an access token [...]

i.e. we must (among other things) make sure that the token is indeed an access token. This is a valid requirement but is different from the one we are currently stating.

I think, this should be:

> Verify that the access token is intended for the service (API). For JWTs this can be achieved by validating the claims 'aud' and 'typ'. If audience is present, it must be validated to match against an allow list for the given API. If the typ claim is present, it should be equal to 'at-jwt'. (see new issue https://github.com/OWASP/ASVS/issues/2363)

**Comment by randomstuff** [comment]
> V13.?? Access tokens

What quite worries me is that some of these requirements are actually valid for a wide range of token outside of access tokens. If we include them in an "access token" section, we would end up repeating them once for each type of token.

**Comment by randomstuff** [comment]
> Verify that access tokens are cryptographically secure, either by using a digital signature or MAC to protect against tampering. If validation fails, the token must be rejected. (see https://github.com/OWASP/ASVS/issues/2184)

The access token can be an opaque random token which would not need to be cryptographically secure using a digital signature or a MAC. If the intent is to disallow this, this should probably be states more explicitly.

**Comment by TobiasAhnoff** [comment]
> The access token can be an opaque random token which would not need to be cryptographically secure using a digital signature or a MAC. If the intent is to disallow this, this should probably be states more explicitly.

Correct, the intent was not to disallow this, an access token can (just as a session identifier) be just a secure random string (without involving OAuth introspection). Perhaps this will include all (I will also update #2184)

> Verify that access tokens are cryptographically secure, either by using a digital signature or MAC to protect against tampering, or being a secure random string (hard to guess and with low collision risk). If validation fails, the token must be rejected. (see https://github.com/OWASP/ASVS/issues/2184)

**Comment by TobiasAhnoff** [comment]
> What quite worries me is that some most of these requirements are actually valid for a wide range of token outside of access tokens. If we include them in an "access token" section, we would end up repeating them once for each type of token.

What kind of tokens are you thinking about? I´m thinking APIs should should only accept some sort of access token (or be part of an application that has sessions). Other kind of tokens: OIDC ID Tokens, OAuth Reference access tokens and OAuth Refresh tokens should be part of (V51).  Is it overlap between this proposed section and V51 you are concerned about (also noted in https://github.com/OWASP/ASVS/issues/2363#issuecomment-2466219113) ? Or something else?

**Comment by elarlang** [comment]
This discussion is getting here not followable and messy.

The original goal was to clean-up 3.5 section. My recommendation to use this issue was in context that if there are more JWT specific issues, then we should put them into 3.5 first and then see do we have enough content for a separate section. Current access token discussion from this point of view is out of scope from this issue, although related.

Whatever is related with precise proposal for any recommended requirement, please have discussion in related opened issue - and if there not any yet, please open new one.

So I recommend to open a new issue for with a proposal of access token related section. But first we need to figure out, do we have enough content for a separate section.

**Comment by elarlang** [comment]
More feedback on the "Access token" proposal.

What problem we solve with this new proposed section? Taking into account how many potential duplication and confusion we create with that.

I like that we have abstract stateless token section. As already [pointed out](https://github.com/OWASP/ASVS/issues/1917#issuecomment-2466491383), access token is may or may not use stateless token.

Even if access token is JWT token, then the main difference for the token type comes when the application already validates the token (expires, signature, issuer, aud etc). General validation is the same procedure for every kind of tokens. "Access token" from the "stateless token" is mostly only the payload, and this can be already really flow-specific, such as OAuth. I don't think we need layers like "stateless token" > "JWT" > "Access token" > "OAuth". I still think stateless token > OAuth is enough.

If it requires more discussion, we can have some video-call. At the moment for me the direction or idea behind is not understandable.

**Comment by elarlang** [comment]
Problems listed here initially have been solved.

The future of the entire section we can discuss here: #2384
